spark-plugins/src/main/java/co/cask/hydrator/plugin/batch/spark/DecisionTreeTrainer.java

```diff
@@ -0,0 +1,197 @@
+    }
+  }
+
+  @Override
+  public void run(SparkExecutionPluginContext context, final JavaRDD<StructuredRecord> input) throws Exception {
+    //categoricalFeaturesInfo : Specifies which features are categorical and how many categorical values each of those
+    //features can take. This is given as a map from feature index to the number of categories for that feature.
+    //Empty categoricalFeaturesInfo indicates all features are continuous.
+    Map<Integer, Integer> categoricalFeaturesInfo = new HashMap<>();
+    final HashMap<String, Map<Object, Long>> categoricalFeaturesMap = new HashMap<>();
+    final String[] fields = config.features.split(",");
+    List<String> categoricalFeaturesList = Arrays.asList(fields);
+
+    for (final String field : fields) {
+      Map<Object, Long> map = input.map(new Function<StructuredRecord, Object>() {
+        @Override
+        public Object call(StructuredRecord structuredRecord) throws Exception {
+          Schema schema = structuredRecord.getSchema().getField(field).getSchema();
+          if (structuredRecord.get(field) == null && !schema.isNullable()) {
+            throw new IllegalArgumentException(String.format("null value found for non-nullable field %s", field));
+          }
+          Schema.Type type = schema.isNullable() ? schema.getNonNullable().getType() : schema.getType();
+          if (!(type.equals(Schema.Type.INT) || type.equals(Schema.Type.DOUBLE) || type.equals(Schema.Type.FLOAT) ||
+            type.equals(Schema.Type.LONG))) {
+            return structuredRecord.get(field);
+          } else {
+            return null;
+          }
+        }
+      }).distinct().zipWithIndex().collectAsMap();
+
+      if (!(map == null || (map.size() == 1 && map.keySet().contains(null)))) {
+        categoricalFeaturesMap.put(field, map);
+      }
+    }
+
+    JavaRDD<LabeledPoint> trainingData = input.map(new Function<StructuredRecord, LabeledPoint>() {
+      @Override
+      public LabeledPoint call(StructuredRecord record) throws Exception {
+        List<Double> doubles = new ArrayList<Double>();
+        Set<String> keys = categoricalFeaturesMap.keySet();
+        for (String field : fields) {
+          if (keys.contains(field)) {
+            doubles.add((categoricalFeaturesMap.get(field).get(record.get(field))).doubleValue());
+          } else {
+            doubles.add(new Double(record.get(field).toString()));
+          }
+        }
+        return new LabeledPoint((Double) record.get(config.predictionField), Vectors.dense(Doubles.toArray(doubles)));
+      }
+    });
+
+    trainingData.cache();
+    Set<String> keys = categoricalFeaturesMap.keySet();
+    for (String key : keys) {
+      categoricalFeaturesInfo.put(categoricalFeaturesList.indexOf(key), categoricalFeaturesMap.get(key).size());
+    }
+
+    config.maxDepth = config.maxDepth == null ? 10 : config.maxDepth;
+    config.maxBins = config.maxBins == null ? 100 : config.maxBins;
```

Comment:
Should the values 10 and 100 be put into static vars? DEFAULT_MAX_DEPTH and DEFAULT_MAX_BINS?